indexing-service/src/main/java/io/druid/indexing/common/actions/SegmentTransactionalInsertAction.java

```diff
@@ -99,28 +103,41 @@ public DataSourceMetadata getEndMetadata()
   @Override
   public SegmentPublishResult perform(Task task, TaskActionToolbox toolbox) throws IOException
   {
-    toolbox.verifyTaskLocks(task, segments);
-
-    final SegmentPublishResult retVal = toolbox.getIndexerMetadataStorageCoordinator().announceHistoricalSegments(
-        segments,
-        startMetadata,
-        endMetadata
-    );
-
-    // Emit metrics
-    final ServiceMetricEvent.Builder metricBuilder = new ServiceMetricEvent.Builder()
-        .setDimension(DruidMetrics.DATASOURCE, task.getDataSource())
-        .setDimension(DruidMetrics.TASK_TYPE, task.getType());
-
-    if (retVal.isSuccess()) {
-      toolbox.getEmitter().emit(metricBuilder.build("segment/txn/success", 1));
-    } else {
-      toolbox.getEmitter().emit(metricBuilder.build("segment/txn/failure", 1));
+    TaskActionPreconditions.checkLockCoversSegments(task, toolbox.getTaskLockbox(), segments);
+
+    final SegmentPublishResult retVal;
+    try {
+      retVal = toolbox.getTaskLockbox().doInCriticalSection(
+          task,
+          segments.stream().map(DataSegment::getInterval).collect(Collectors.toList()),
+          () -> toolbox.getIndexerMetadataStorageCoordinator().announceHistoricalSegments(
```

Comment:
This comment makes me wonder, do we need critical sections at all for locks?

Could the cases that require them all be implemented using metadata store transactions?

Reply:
```doInCriticalSection``` guarantees the lock validation and a critical action are done atomically. I thought the metadata store transactions are used for storing/retrieving segment information to/from the metadata storage, and TaskLockBox is responsible for managing task locks. It means, lock validity should be checked by TaskLockBox. 

Do you think locks can be checked inside metadata storage transactions as well? I think it might be possible if all locks are stored in only metadata store not in memory. Also, I wonder why locks are stored in memory as well as metadata store. Is this for performance?

Reply:
Ah, I see. You're right. I almost forgot that lock checking is important, since for insert-only tasks the design does not depend on lock checking for correctness. But it does matter for mixing insert tasks with reindexing or updating tasks.

> Do you think locks can be checked inside metadata storage transactions as well? I think it might be possible if all locks are stored in only metadata store not in memory. Also, I wonder why locks are stored in memory as well as metadata store. Is this for performance?

Hmm, I think it's worth exploring...

The idea behind storing locks in memory was indeed performance. However with the really fat synchronization I'm not sure how much benefit we get beyond storing them solely in the metadata store. I haven't evaluated the performance of metadata-store-only vs. the current memory+metadata system. It would surely make things simpler to do metadata-store-only, all else being equal.

Reply:
Sounds good. I'll do some benchmark in a follow-up pr.